File: trunk/AStyleDev/src-jx/Example2.java
// Example2.java

/*
* Example2 uses the AStyleInterface class to call the Artistic Style formatter
* (AStyleMain) to format source files in a 'test-j' directory.
 */

import java.io.*;

public class Example2
{   public static void main(String[] args)
    {   // files to pass to AStyle
        String fileName[] =  { "../test-j/JSBeautifier.java",
                               "../test-j/JSFormatter.java" ,
                               "../test-j/JSLineBreaker.java" ,
                             };

        // create an object
        AStyleInterface astyle = new AStyleInterface();

        // set test values for the options
        astyle.setTestOptions();

        // get Artistic Style version
        // does not need to terminate on an error
        String version = astyle.getVersion();
        if (version.length() != 0)
            System.out.println("AStyle Version " + version);

        // process the files
        for (int i = 0; i < fileName.length; i++)
        {   // get the text to format
            String textIn = getText(fileName[i]);

            // call the Artistic Style formatting function
            // does not need to terminate on an error
            String textOut = astyle.formatSource(textIn, fileName[i]);
            if (textOut.length() == 0)
            {   System.out.println("Cannot format "  + fileName[i]);
                continue;
            }

            // return the formatted text
            System.out.println("formatted " + fileName[i]);
            setText(textOut, fileName[i]);
        }
        return;
    }

    /*
    *  Error message function for this example.
    */
    private static void error(String why, String what)
    {   System.out.println(why + ' ' + what);
        System.out.println("The program has terminated!");
        System.exit(1);
    }

    /*
    *  Get the text to be formatted.
    *  Usually the text would be obtained from an edit control.
    */
    private static String getText(String fileName)
    {   // create input buffers
        File inFile = new File(fileName);
        int readSize = 1024;
        StringBuffer bufferIn = new StringBuffer(readSize);
        char fileIn[] = new char[readSize];

        // read file data
        try
        {   BufferedReader in =
                new BufferedReader(new FileReader(inFile));
            // use read to preserve the current line endings
            int charsIn = in.read(fileIn, 0, readSize);
            while (charsIn != -1)
            {   bufferIn.append(fileIn, 0, charsIn);
                charsIn = in.read(fileIn, 0, readSize);
            }
            in.close();
        }
        catch (Exception e)
        {   if (e instanceof FileNotFoundException)
                error("Could not open input file", fileName);
            else if (e instanceof IOException)
                error("Error reading file", fileName);
            else
                error(e.getMessage(), fileName);
        }

        return bufferIn.toString();
    }

    /*
    *  Return the formatted text.
    *  Usually the text would be returned to an edit control.
    */
    private static void setText(String textOut, String fileName)
    {   // create a backup file
        String origFileName = fileName +  ".orig";
        File origFile = new File(origFileName);
        File outFile = new File(fileName);
        origFile.delete();                  // remove a pre-existing file
        if (!outFile.renameTo(origFile))
            error("Could not create backup file", origFileName);

        // write the output file - same name as input
        try
        {   BufferedWriter out =
                new BufferedWriter(new FileWriter(fileName));
            out.write(textOut, 0, textOut.length());
            out.close();
        }
        catch (IOException e)
        {   error("Could not write to output", fileName);
        }
    }

}   // class Example2

File: trunk/AStyleDev/src-cx/Example2.cpp
// Example2.cpp

/* This program calls the Artistic Style GUI formatter (AStyleMain)
 * to format the astyle source files in a test-c directory.
 */

#include "AStyleInterface.h"

#include <string.h>         // need both string and string.h for GCC
#include <stdlib.h>
#include <fstream>
#include <iostream>

// function declarations
void error(const char* why, const char* what = "");
string getProjectDirectory(string& subPath);
char* getText(string& filePath);
void  setText(const char* textOut, string& filePathStr);
void setOptionValues(AStyleInterface& as);


int main(int, char**)
{   // options to pass to AStyle
    string fileName[] = { "AStyleDev/test-c/ASBeautifier.cpp",
                          "AStyleDev/test-c/ASFormatter.cpp" ,
                          "AStyleDev/test-c/astyle.h"
                        };
    size_t arraySize = sizeof(fileName) / sizeof(fileName[0]);

    // create an object
    AStyleInterface astyle;

    // set test values for the options
    astyle.setTestOptions();

    // get Artistic Style version
    char* version = AStyleGetVersion();
    cout << "Example2 C++ - AStyle " << version << endl;

    // process the input files
    for (size_t i = 0; i < arraySize; i++)
    {   // get the text to format
        string filePath = getProjectDirectory(fileName[i]);
        char* textIn = getText(filePath);

        // call the Artistic Style formatting function
        char* textOut = astyle.formatSource(textIn, filePath);

        // NULL pointer is an error - restore the original file
        // an error message has been displayed by the error handler
        if (textOut == NULL)
        {   cout << "Cannot format " << filePath << endl;
            return 0;
        }

        // return the formatted text
        cout << "Formatted " << fileName[i] << endl;
        setText(textOut, filePath);

        // must delete the temporary buffers
        delete [] textIn;
        delete [] textOut;
    }
#ifdef __MINGW32__
    system("pause");
#endif
}

// Error message function for this example
void error(const char* why, const char* what)
{   cout << why << ' ' << what << endl;
    cout << "The program has terminated!" << endl;
    exit(1);
}

// Prepend the project directory to the subpath
// This may need to be changed for your directory structure
string getProjectDirectory(string& subPath)
{
#ifdef _WIN32
    char* homeDirectory = getenv("USERPROFILE");
#else
    char* homeDirectory = getenv("HOME");
#endif
    if (!homeDirectory)
        error("Cannot find HOME directory");
    string projectPath = string(homeDirectory) + "/Projects/" + subPath;
    return projectPath;
}

// Get the text to be formatted
// Usually the text would be obtained from an edit control
char* getText(string& filePath)
{   // open input file
    ifstream in(filePath.c_str());
    if (!in)
        error("Cannot open input file", filePath.c_str());

    // get length of buffer
    const int bufferSizeIn = 131072;     // 128 KB

    // allocate memory
    char* bufferIn = new(nothrow) char [bufferSizeIn];
    if (bufferIn == NULL)
    {   in.close();
        error("Memory allocation failure on input");
    }

    // read data as a block
    in.read(bufferIn, bufferSizeIn);
    // get actual size - must be smaller than buffer size
    int textSizeIn = static_cast<int>(in.gcount());
    if (textSizeIn == bufferSizeIn)
    {   in.close();
        error("Read buffer is too small");
    }
    bufferIn[textSizeIn] = '\0';
    in.close();

    return bufferIn;
}

// Return the formatted text
// Usually the text would be returned to an edit control
void setText(const char* textOut, string& filePathStr)
{   // create a backup file
    const char* filePath = filePathStr.c_str();
    string origfilePathStr = filePathStr + ".orig";
    const char* origfilePath = origfilePathStr.c_str();
    remove(origfilePath);              // remove a pre-existing file

    if (rename(filePath, origfilePath) < 0)
        error("Cannot open input file", filePath);

    // open the output file
    ofstream out(filePath);
    if (!out)
        error("Cannot open output file", filePath);

    // write the text
    int textSizeOut = strlen(textOut);
    out.write(textOut, textSizeOut);
    out.close();
}
